<!DOCTYPE html>
        <html>
        <head>
            <meta charset="UTF-8">
            <title>Xenium Ranger</title>
            <style>
/* From extension vscode.github */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.vscode-dark img[src$=\#gh-light-mode-only],
.vscode-light img[src$=\#gh-dark-mode-only],
.vscode-high-contrast:not(.vscode-high-contrast-light) img[src$=\#gh-light-mode-only],
.vscode-high-contrast-light img[src$=\#gh-dark-mode-only] {
	display: none;
}

</style>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex/dist/katex.min.css">
<link href="https://cdn.jsdelivr.net/npm/katex-copytex@latest/dist/katex-copytex.min.css" rel="stylesheet" type="text/css">
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/markdown.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/highlight.css">
<style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif;
                font-size: 14px;
                line-height: 1.6;
            }
        </style>
        <style>
.task-list-item {
    list-style-type: none;
}

.task-list-item-checkbox {
    margin-left: -20px;
    vertical-align: middle;
    pointer-events: none;
}
</style>
<style>
:root {
  --color-note: #0969da;
  --color-tip: #1a7f37;
  --color-warning: #9a6700;
  --color-severe: #bc4c00;
  --color-caution: #d1242f;
  --color-important: #8250df;
}

</style>
<style>
@media (prefers-color-scheme: dark) {
  :root {
    --color-note: #2f81f7;
    --color-tip: #3fb950;
    --color-warning: #d29922;
    --color-severe: #db6d28;
    --color-caution: #f85149;
    --color-important: #a371f7;
  }
}

</style>
<style>
.markdown-alert {
  padding: 0.5rem 1rem;
  margin-bottom: 16px;
  color: inherit;
  border-left: .25em solid #888;
}

.markdown-alert>:first-child {
  margin-top: 0
}

.markdown-alert>:last-child {
  margin-bottom: 0
}

.markdown-alert .markdown-alert-title {
  display: flex;
  font-weight: 500;
  align-items: center;
  line-height: 1
}

.markdown-alert .markdown-alert-title .octicon {
  margin-right: 0.5rem;
  display: inline-block;
  overflow: visible !important;
  vertical-align: text-bottom;
  fill: currentColor;
}

.markdown-alert.markdown-alert-note {
  border-left-color: var(--color-note);
}

.markdown-alert.markdown-alert-note .markdown-alert-title {
  color: var(--color-note);
}

.markdown-alert.markdown-alert-important {
  border-left-color: var(--color-important);
}

.markdown-alert.markdown-alert-important .markdown-alert-title {
  color: var(--color-important);
}

.markdown-alert.markdown-alert-warning {
  border-left-color: var(--color-warning);
}

.markdown-alert.markdown-alert-warning .markdown-alert-title {
  color: var(--color-warning);
}

.markdown-alert.markdown-alert-tip {
  border-left-color: var(--color-tip);
}

.markdown-alert.markdown-alert-tip .markdown-alert-title {
  color: var(--color-tip);
}

.markdown-alert.markdown-alert-caution {
  border-left-color: var(--color-caution);
}

.markdown-alert.markdown-alert-caution .markdown-alert-title {
  color: var(--color-caution);
}

</style>
        
        </head>
        <body class="vscode-body vscode-light">
            <h1 id="xenium-ranger">Xenium Ranger</h1>
<h2 id="xenium-ranger-is-a-set-of-analysis-pipelines-that-process-xenium-in-situ-gene-expression-data-it-includes-the-following-pipelines">Xenium Ranger is a set of analysis pipelines that process Xenium In Situ Gene Expression data. It includes the following pipelines:</h2>
<ul>
<li>Relabel transcripts with <code>xeniumranger relabel</code></li>
<li>Resegment cells with latest 10x segmentation algorithms with <code>xeniumranger resegment</code></li>
<li>Import your own segmentation data to assign transcripts to cells with <code>xeniumranger import-segmentation</code></li>
<li>Rename region or cassette names with <code>xeniumranger rename</code></li>
</ul>
<h2 id="download--install-xeniumranger">Download &amp; install Xeniumranger</h2>
<pre><code class="language-shell">cd /opt
sudo mkdir Xeniumranger
cd Xeniumranger
curl -o xeniumranger-2.0.1.tar.gz &quot;https://cf.10xgenomics.com/releases/xeniumranger/xeniumranger-2.0.1.tar.gz?Expires=1725300474&amp;Key-Pair-Id=APKAI7S6A5RYOXBWRPDA&amp;Signature=BITYA48Bx-M4Md6K-RFIuAPhKdeZbpNgq6izDUNPLCO0OrPT5wY9E6tK8w4xK7nX2ER2YrN0faTrb13IjNztjZ6uYHfa~dfAZTlC2S361tWVxhG9l192jn9cniW7YhNPhVTCXl2Evkpu5U~4pVTV4O4SaKwYRwjNYnPmXCG6fMvC-Y6fbzQEsU3JrDGloxoYwU22vtD789yCnfqlCGaHU128-ypJ5loZDwa7wMalKXFXmuPMP9Q8vOCtBVof~MubPFaHN6ESy59WNCckGyxFm1ABRqTYQdYnjaZG-EhFqSLjtIZdMNPsLQ4X4yZP0QYPYF~SDIc5ciNeN2nAASijsA__&quot;
tar -xzvf xeniumranger-2.0.1.tar.gz
export PATH=/opt/Xeniumranger/xeniumranger-xenium2.0.1:$PATH
</code></pre>
<h2 id="optional-relabel-decoded-transcripts">(Optional) Relabel Decoded Transcripts</h2>
<p>The <code>relabel</code> pipeline allows you to change the gene labels applied to decoded transcripts. If the wrong gene_panel.json file was selected on the Xenium Analyzer, you can use this pipeline to correct the gene labels after the instrument run completes.</p>
<pre><code class="language-shell">xeniumranger relabel --id=relabel-demo \
                     --xenium-bundle=/path/to/xenium/files \
                     --panel=/opt/xeniumranger-xenium2.0/lib/json/definitions/panel_designer/panels/xenium_human_brain_gene_expression_panel.json \
                     --localcores=32 \
                     --localmem=128
</code></pre>
<h2 id="optional-relabel-decoded-transcripts-1">(Optional) Relabel Decoded Transcripts</h2>
<p>The <code>rename</code> pipeline allows you to change the sample <code>region_name</code> and <code>cassette_name</code> throughout all the Xenium Onboard Analysis output files that contain this information.</p>
<pre><code class="language-shell">xeniumranger rename --id=rename-demo \
                     --xenium-bundle=/path/to/xenium/files \
                     --region-name=&quot;replicate 1&quot; \
                     --cassette-name=&quot;cassette 1&quot;
</code></pre>
<h2 id="optional-rerun-xenium-onboard-analysis-segmentation">(Optional) Rerun Xenium Onboard Analysis Segmentation</h2>
<p>The <code>resegment</code> pipeline allows you to generate a new segmentation of the morphology image space</p>
<pre><code class="language-shell">xeniumranger resegment --id=demo \
                       --xenium-bundle=/path/to/xenium/files \
                       --localcores=32 \
                       --localmem=128
</code></pre>
<h2 id="import-nucleus-and-cell-segmentation-results">Import Nucleus and Cell Segmentation Results</h2>
<p>The <code>import-segmentation</code> pipeline allows you to specify 2D nuclei and/or cell segmentation results to use for assigning transcripts to cells and recalculate all Xenium Onboard Analysis (XOA) outputs that depend on segmentation.</p>
<pre><code class="language-shell"><span class="hljs-meta prompt_"># </span><span class="language-bash">Only change nuclear expansion</span>
xeniumranger import-segmentation --id=demo \
                                 --xenium-bundle=/path/to/xenium/files \
                                 --nuclei=/path/to/cells.zarr.zip \
                                 --localcores=32 \
                                 --localmem=128
                    
<span class="hljs-meta prompt_"># </span><span class="language-bash">Nucleus-only count matrix</span>
xeniumranger import-segmentation --id=demo \
                                 --xenium-bundle=/path/to/xenium/files \
                                 --nuclei=/path/to/cells.zarr.zip \
                                 --expansion-distance=0 \
                                 --localcores=32 \
                                 --localmem=128
<span class="hljs-meta prompt_">
# </span><span class="language-bash">Polygon input</span>
xeniumranger import-segmentation --id=polygon-demo \
                                 --xenium-bundle=/path/to/xenium/files \
                                 --nuclei=polygon.geojson \
                                 --cells=polygon.geojson \
                                 --units=pixels \
                                 --localcores=32 \
                                 --localmem=128
                    
<span class="hljs-meta prompt_"># </span><span class="language-bash">Mask input</span>
xeniumranger import-segmentation --id=mask-demo-nuclei \
                                 --xenium-bundle=/path/to/xenium/files \
                                 --nuclei=nuclei.tif \
                                 --cells=cells.npy \
                                 --localcores=32 \
                                 --localmem=128
<span class="hljs-meta prompt_">
# </span><span class="language-bash">Transcript assignment input</span>
xeniumranger import-segmentation --id=baysor-demo \
                                 --xenium-bundle=/path/to/xenium/files \
                                 --transcript-assignment=segmentation.csv \
                                 --viz-polygons=segmentation_polygons.json \
                                 --units=microns \
                                 --localcores=32 \
                                 --localmem=128
<span class="hljs-meta prompt_">
# </span><span class="language-bash">Transformation matrix input</span>
xeniumranger import-segmentation --id=image-transform-demo \
                                 --xenium-bundle=/path/to/xenium/files \
                                 --coordinate-transform=demo_imagealignment.csv \
                                 --nuclei=qupath.geojson \
                                 --cells=qupath.geojson \
                                 --units=pixels \
                                 --localcores=32 \
                                 --localmem=128
</code></pre>
<h2 id="output">Output</h2>
<ul>
<li>Analysis summary: check metrics and plots on the Cell Segmentation tab. There are new metrics for imported segmentation results (present in <code>analysis_summary.html</code> and <code>metrics_summary.csv</code>).</li>
<li>All secondary analysis files</li>
<li>All cell, nuclei, and transcript files</li>
</ul>

            <script async src="https://cdn.jsdelivr.net/npm/katex-copytex@latest/dist/katex-copytex.min.js"></script>
            
        </body>
        </html>